include/quic/datagram.hpp

```diff
@@ -1,16 +1,71 @@
 #pragma once
 
+#include "udp.hpp"
 #include "utils.hpp"
 
 namespace oxen::quic
 {
+    extern std::atomic<bool> enable_test_features;
+    extern std::atomic<int> test_counter;
+
     class IOChannel;
     class Connection;
     class Endpoint;
     class Stream;
 
     // IO callbacks
-    using dgram_data_callback = std::function<void(bstring_view)>;
+    using dgram_data_callback = std::function<void(bstring)>;
+
+    using dgram_buffer = std::deque<std::pair<uint16_t, std::pair<bstring_view, std::shared_ptr<void>>>>;
+
+    struct prepared_datagram
+    {
+        uint64_t id;                  // internal ID for ngtcp2
+        std::array<uint8_t, 2> dgid;  // optional transmitted ID buffer (for packet splitting)
+        std::array<ngtcp2_vec, 2> bufs;
+        size_t bufs_len;  // either 1 or 2 depending on how much of data is populated
+
+        const ngtcp2_vec* data() const { return bufs.data(); }
+        size_t size() const { return bufs_len; }
+    };
+
+    struct received_datagram
+    {
+        uint16_t id{0};
+        bstring data{};
+        // -1 = first half, 1 = second half
+        int part{0};
```

Comment:
Any reason not just to use a bool (maybe `first_part`) for this?

Also, whether bool or not, this would be slightly better placed between `id` and `data` to make the struct layout *slightly* smaller: `bstring` most likely uses size_t-alignment (i.e. 8-byte alignment), so if you do it this way you have a struct containing data like this: `[u16][6 bytes padding][string, probably 24 bytes][int][4 bytes padding]`, whereas if you rearrange the smaller types together you will get: `[u16][2 padding][int][24-byte-string]` and make the struct 8 bytes smaller.  Not a huge deal, but also easy to change.